### DISCUSSION TITLE
[Q&A] openai 的接口问什么要从 vercel 服务端绕一圈?

### DISCUSSION THREAD

**Discussion by yh4922** [lobehub/lobehub#728] [General Question] [answered]
### 🧐 问题描述 | Proposed Solution

能直接从浏览器发送请求应该会好点吧  类似  [ChatGPT-Next-Web](https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web) 的方式

**Comment by arvinxx** [comment, marked as answer]
两方面。一方面是服务端可以隐藏key。如果直接在前端发请求，那么apikey会暴露，这个是很多人不希望的。

另一方面，只要服务器位置得当，就可以做到直连openai，不用魔法或者三方代理。

最后， chat next 应该默认也是用的服务端请求。

**Reply by suwubee** [reply]
1，现在的一样暴露在header里（自填api url）
2，目前大多数都是用的转发url，也没有跨域问题。
3，openai封ip特别厉害，就我个人而言也会自建转发。
4，带来的速度损失得不偿失
建议加开关或者依然由前端直接请求。也减轻程序的压力。

**Reply by arvinxx** [reply]
> 现在的一样暴露在header里（自填api url）

我是指非手动填入 api key 的方式。如果自部署时你在环境变量中填写了api key，终端用户就可以直接使用。由于是服务端请求，就不会出现需要把apikey发到前端，再进行前端请求的情况。这种模式是 LobeChat 用户主要的使用方式。

如果是前端填写apikey进行使用的方式，那么的确可能前端请求更理想。但我也说了，这是个研发成本的问题。在用户诉求更强的场景下可能会做，但不是现在。

**Reply by Yidadaa** [reply]
Next Web 的逻辑是这样的：
- 默认情况下走 `api/openai`，也就是 vercel 中转，经由 vercel 服务器发往 BASE_URL 环境变量指向的地址；
- 如果用户修改 `api/openai` 为其他地址，请求就不会走 vercel 中转，而是直接发往该地址。

**Reply by se7enmuting** [reply]
小白，上面没太看懂，想最终简单确认下：
我通过docker部署在国外服务器，docker环境变量里配置了 OPENAI_API_KEY 和 OPENAI_PROXY_URL（第三方转发的url），这种配置下，我在国内访问（直接用，不在前端上填写API_KEY），请问还会绕 vercel 服务端？这种情况 apikey 的请求是在我docker的服务器端发送的，还是在我本地前端发送的？

**Reply by arvinxx** [reply]
从你的 docker 服务端发送的。

[a comment by an automated account is not shown]

**Comment by yh4922** [comment]
如果实在前端输入key的话 也要经过服务端中转么

**Reply by arvinxx** [reply]
对。原因有两点：

1、针对两种场景做两套实现的成本有点高；
2、其他三方api代理不一定允许跨域访问( OpenAI 是支持的 )；

**Reply by arvinxx** [reply]
当然在前端做也不是不行。也有潜在的两点好处：

1. 可以支持向本地起来的 LLM 服务发送请求；
2. 避免本地填写 api-key 泄漏的风险 （refs: https://github.com/lobehub/lobe-chat/discussions/561 ）

**Reply by tisfeng** [reply]
请问是否可以提供一个中转选项？这个值可以默认 YES，有时候我不希望接口中转 🤔

**Reply by arvinxx** [reply]
@tisfeng 后续会做 https://github.com/lobehub/lobe-chat/discussions/1257 。 可以关注这个进展

**Comment by arvinxx** [comment]
@yh4922 @tisfeng @suwubee 已在此 PR https://github.com/lobehub/lobe-chat/pull/2041 实现了客户端请求

**Reply by tisfeng** [reply]
你好，我看最新 vercel 部署版本模型设置页面已经支持 Ollama，但测试使用就一直失败，如何解决？

> 已在此 PR https://github.com/lobehub/lobe-chat/pull/2041 实现了客户端请求

请问现在默认就是客户端请求连接吗，还是说需要额外设置？

我是 Mac 环境，确认本地的 Ollama 接口没有问题。

<img width="1673" alt="image" src="https://github.com/lobehub/lobe-chat/assets/25194972/e01f5562-5f38-428a-9eee-2f5c9780a953">

<img width="1236" alt="image" src="https://github.com/lobehub/lobe-chat/assets/25194972/05bea311-36fd-4ef2-9d7b-f5f0e716cd6c">

**Reply by arvinxx** [reply]
> 请问现在默认就是客户端请求连接吗，还是说需要额外设置？

只有填写了代理（proxy url）之后，才会默认开启客户端请求。你填写之后应该能看到一个请求控制的参数

**Reply by tisfeng** [reply]
试过，还是不行

<img width="1673" alt="image" src="https://github.com/lobehub/lobe-chat/assets/25194972/fca0b35f-e2be-4400-a1cc-c25088877b9c">

**Reply by tisfeng** [reply]
不理解，ollama 这种不是应该默认开启直连吗，ollama 本地请求还能走代理？

**Reply by tisfeng** [reply]
> 填入 Ollama 接口代理地址，本地未额外指定可留空

就算是出于某种特殊原因需要默认关闭直连，那也应该将这个 UI 选项显示出来，不然【本地未额外指定可留空】这个描述太误导用户了 😓

<img width="1055" alt="image" src="https://github.com/lobehub/lobe-chat/assets/25194972/3371e1a8-c2c5-43a3-abef-989bc9fa5b50">

**Reply by arvinxx** [reply]
> 不理解，ollama 这种不是应该默认开启直连吗，ollama 本地请求还能走代理？

用户的使用场景是多样的…我做 ollama 的时候都没想过会有用户在服务端（自己的 linux）机器上部署 ollama 和 lobe-chat 的 docker-compose。那在这种场景下的确仍然需要一个本地请求的地址。

另外 ollama 这块还没做优化。包括提示这些也都是按之前 docker 方案来的，所以在本地直连场景下还是有点问题，后续会继续优化。

还有一点是，纯本地浏览器请求的场景，ollama 需要配置允许跨域，不知道你有没有做这个配置：https://lobehub.com/zh/docs/self-hosting/examples/ollama#%E9%9D%9E%E6%9C%AC%E5%9C%B0%E8%AE%BF%E9%97%AE-ollama

**Reply by arvinxx** [reply]
ollama 使用这块有问题建议单独起一个讨论吧。不再这里继续聊了

**Reply by tisfeng** [reply]
ok，明白了。
